File: LeetCode/Study Plan/Practice Problems/Easy/Leaf Similar Trees/Cpp/leaf_similar_trees.cc
#include <iostream>
#include <vector>

#define PB push_back

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

typedef std::vector<int> vi;

void traverse(TreeNode *r, vi &v) {
    if (r == NULL) return;
    if (r->left == NULL && r->right == NULL) v.PB(r->val);
    traverse(r->left, v);
    traverse(r->right, v);
}

bool leaf_similar(TreeNode* r1, TreeNode* r2) {
    vi v1, v2;
    traverse(r1, v1);
    traverse(r2, v2);
    if (v1.size() != v2.size()) return false;
    for (int i=0; i<v1.size(); i++) if (v1[i] != v2[i]) return false;
    return true;
}

int main () {
    TreeNode *r1 = new TreeNode(3);
    r1->left = new TreeNode(1);
    r1->right = new TreeNode(5);
    TreeNode *r2 = new TreeNode(2);
    r2->left = new TreeNode(1);
    r2->right = new TreeNode(5);
    std::cout<<leaf_similar(r1, r2)<<"\n";
}

File: LeetCode/Easy/Leaf Similar Trees/Java/leafSimilarTrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> leaf1 = new ArrayList<>();
        List<Integer> leaf2 = new ArrayList<>();
        recur(leaf1, root1);
        recur(leaf2, root2);
        return leaf1.equals(leaf2);
    }

    private static void recur(List<Integer> leaf, TreeNode root) {
        if (root == null) return;
        if (root.left == null && root.right == null) leaf.add(root.val);
        recur(leaf, root.left);
        recur(leaf, root.right);
    } 
}
